src/client/Microsoft.Identity.Client/Internal/Requests/ClientCredentialRequest.cs

```diff
@@ -117,10 +117,36 @@ protected override async Task<AuthenticationResult> ExecuteAsync(CancellationTok
         private async Task<AuthenticationResult> FetchNewAccessTokenAsync(CancellationToken cancellationToken)
         {
             await ResolveAuthorityAsync().ConfigureAwait(false);
-            var msalTokenResponse = await SendTokenRequestAsync(GetBodyParameters(), cancellationToken).ConfigureAwait(false);
+            MsalTokenResponse msalTokenResponse;
+            if (ServiceBundle.Config.AppTokenProviderDelegate != null)
+            {
+                msalTokenResponse = await SendTokenRequestToProviderAsync(cancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                msalTokenResponse = await SendTokenRequestAsync(GetBodyParameters(), cancellationToken).ConfigureAwait(false);
+            }
+
             return await CacheTokenResponseAndCreateAuthenticationResultAsync(msalTokenResponse).ConfigureAwait(false);
         }
 
+
+        private async Task<MsalTokenResponse> SendTokenRequestToProviderAsync(CancellationToken cancellationToken)
+        {
+            AppTokenProviderParameters appTokenProviderParameters = new AppTokenProviderParameters();
+            appTokenProviderParameters.Scopes = GetOverriddenScopes(AuthenticationRequestParameters.Scope);
+            appTokenProviderParameters.CorrelationId = AuthenticationRequestParameters.RequestContext.CorrelationId.ToString();
+            appTokenProviderParameters.Claims = AuthenticationRequestParameters.Claims;
+            appTokenProviderParameters.TenantId = AuthenticationRequestParameters.Authority.TenantId;
+            appTokenProviderParameters.CancellationToken = cancellationToken;
+
+            ExternalTokenResult externalToken = await Task.Run(() => ServiceBundle.Config.AppTokenProviderDelegate(appTokenProviderParameters)).ConfigureAwait(false);
```

Comment:
Again, Task.Run will put this logic on a different thread and run it, then wait for it. while this works, it's much simpler to do

```csharp
await ServiceBundle.Config.AppTokenProviderDelegate(appTokenProviderParameters);
```

Reply:
Yes, I missed this one.